app/src/behaviors/behavior_macro.c

```diff
@@ -0,0 +1,192 @@
+    } else if (IS_TAP_TIME(binding->behavior_dev)) {
+        state->tap_ms = binding->param1;
+        LOG_DBG("macro tap time set: %d", state->tap_ms);
+    } else if (IS_WAIT_TIME(binding->behavior_dev)) {
+        state->wait_ms = binding->param1;
+        LOG_DBG("macro wait time set: %d", state->wait_ms);
+    } else {
+        return false;
+    }
+
+    return true;
+}
+
+static int behavior_macro_init(const struct device *dev) {
+    const struct behavior_macro_config *cfg = dev->config;
+    struct behavior_macro_state *state = dev->data;
+    state->last_press_binding = cfg->count - 1;
+    state->release_state.start_index = cfg->count;
+    state->release_state.end_index = cfg->count - 1;
+
+    LOG_DBG("Precalculate initial release state:");
+    for (int i = 0; i < cfg->count; i++) {
+        if (handle_control_binding(&state->release_state, &cfg->bindings[i])) {
+            // Updated state used for initial state on release.
+        } else if (IS_PAUSE(cfg->bindings[i].behavior_dev)) {
+            state->release_state.start_index = i + 1;
+            state->release_state.end_index = cfg->count - 1;
+            state->last_press_binding = i - 1;
+            LOG_DBG("Release will resume at %d", state->release_state.start_index);
+            break;
+        } else {
+            // Ignore regular invokable bindings
+        }
+    }
+
+    return 0;
+};
+
+static void queue_macro(uint32_t position, const struct zmk_behavior_binding (*bindings)[],
+                        struct behavior_macro_trigger_state trigger_state) {
+    LOG_DBG("Iterating macro bindings from %d-%d", trigger_state.start_index,
+            trigger_state.end_index);
+    for (int i = trigger_state.start_index; i <= trigger_state.end_index; i++) {
+        if (!handle_control_binding(&trigger_state, &(*bindings)[i])) {
+            switch (trigger_state.mode) {
+            case MACRO_MODE_TAP:
+                zmk_behavior_queue_add(position, (*bindings)[i], true, trigger_state.tap_ms);
+                zmk_behavior_queue_add(position, (*bindings)[i], false, trigger_state.wait_ms);
+                break;
+            case MACRO_MODE_PRESS:
+                zmk_behavior_queue_add(position, (*bindings)[i], true, trigger_state.wait_ms);
+                break;
+            case MACRO_MODE_RELEASE:
+                zmk_behavior_queue_add(position, (*bindings)[i], false, trigger_state.wait_ms);
+                break;
+            default:
+                LOG_ERR("Unknown macro mode: %d", trigger_state.mode);
+                break;
+            }
+        }
+    }
+}
+
+static int on_macro_binding_pressed(struct zmk_behavior_binding *binding,
+                                    struct zmk_behavior_binding_event event) {
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_macro_config *cfg = dev->config;
+    struct behavior_macro_state *state = dev->data;
+
+    LOG_DBG("");
```

Comment:
Are these empty logging statements intentional?

Reply:
I sometimes find them useful for certain scenarios to just see entry into functions happening. Removed.